CAIPs/caip-x.md

```diff
@@ -0,0 +1,110 @@
+
+Each wallet extension must implement a mechanism to announce its extension ID along with the provider info to a dApp. This should be done using a JavaScript event emitted by the wallet and listened for by the dApp. The event should contain the `ProviderInfo` object as its payload.
+
+**Modification to dApp Listening Mechanisms**:
+
+dApps must modify their initialization routines to listen for the wallet's `ProviderInfo` announcement event. Upon capturing the event, the dApp can store the extension ID and use it to establish a connection using the `externally_connectable` API.
+
+
+**Request Provider Mechanism**:
+
+To ensure dApps can discover wallets even if they start listening after wallets have emitted their initial events, a request mechanism is essential.
+
+- **Active Request Capability**: Wallets must listen for a specific event from dApps that requests the announcement of available wallet providers. Upon receiving this event, wallets should re-announce their presence.
+
+- **Event Handling**: Wallets should listen for a `requestWalletExtensionId` event and respond by re-dispatching their provider information using the `walletExtensionIdAnnouncement` event. This ensures that any dApp can discover all available wallets, regardless of the script execution order.
+
+### Example Implementation
+
+- **Wallets**:
+
+```javascript
+function announceExtensionId() {
+  const providerInfo = {
+    name: 'Example Wallet',
+    icon: 'data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg"/>',
+    rdns: 'com.example.wallet',
+    extensionId: 'your_extension_id_here'
+  };
+  document.dispatchEvent(new CustomEvent('walletExtensionIdAnnouncement', {
+    detail: Object.freeze(providerInfo)
+  }));
+}
+
+announceExtensionId();
+
+window.addEventListener('requestWalletExtensionId', () => {
+  announceExtensionId();
+});
+```
+
+- **dApps**:
+
+```javascript
+window.addEventListener('walletExtensionIdAnnouncement', function(e) {
+  const providerInfo = e.detail;
+  // Use providerInfo.extensionId to establish communication
+});
+
+window.dispatchEvent(new Event("requestWalletExtensionId"));
+```
+
+**Content Script Utilization**:
+While this CAIP significantly advances the way wallets interact with dApps through the `externally_connectable` API, it still necessitates the use of a content script. This script, however, requires fewer permissions compared to traditional methods and can be injected synchronously to facilitate the initial communication process. This adjustment is necessary to bridge the current browser limitations and set the stage for future enhancements.
+
+- **Synchronous Injection**: The content script can be injected synchronously with the page load, ensuring that the wallet's extension ID is available as soon as the dApp begins its initialization process. This method avoids the delays and potential conflicts associated with asynchronous script loading, leading to a smoother and more predictable interaction pattern.
```

Comment:
i'm not entirely sure about this point. Technically the dApp will receives the extension IDs asynchronously via events. Even if the dApp is initialized and listening before the wallets have announced themselves, the dApp still must wait for the announcement events to come in before it can proceed.

Also, synchronous injection isn't the more ideal strategy. Supporting asynchronous is as it does not block the execution loop. I don't think we need to mention how the script is injected, only that the injected script is magnitudes smaller, and thus quicker to parse and run, than the previous model of injecting an entire provider object

Reply:
>  the dApp still must wait for the announcement events to come in before it can proceed.

True, but if we can change [run_at](https://developer.chrome.com/docs/extensions/reference/api/extensionTypes#type-RunAt) from `document_start` to `document_end` we block less of every site's DOM construction process for  parsing/execution of our script. Its a fair point that this is much lighter weight script and that would improve performance on its own as compared to the current situation (if other content scripts were removed). Perhaps the performance difference here is negligent and not worth mentioning